Recover deleted draft submissions by soft-deleting them when updating the form draft

ktuite opened the pull request:
Closes https://github.com/getodk/central/issues/749

WIP while adding more tests and thinking about encrypted forms.

<!-- 
Thank you for contributing to ODK Central!

Before sending this PR, please read
https://github.com/getodk/central-backend/blob/master/CONTRIBUTING.md
-->

#### What has been done to verify that this works as intended?

#### Why is this the best possible solution? Were any other approaches considered?

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

#### Does this change require updates to the API documentation? If so, please update docs/api.yaml as part of this PR.

#### Before submitting this PR, please make sure you have:

- [ ] run `make test` and confirmed all checks still pass OR confirm CircleCI build passes
- [ ] verified that any code from external sources are properly credited in comments or that everything is internally sourced

matthew-white commented on lib/resources/forms.js:
```diff
@@ -137,7 +137,7 @@ module.exports = (service, endpoint) => {
           : getPartial(Forms, request, project, Keys)))
         .then((partial) => Promise.all([
           Forms.createVersion(partial, form, false),
-          Submissions.clearDraftSubmissions(form.id)
```
Are we now able to remove the `clearDraftSubmissions()` method entirely?

matthew-white commented on lib/resources/forms.js:
```diff
@@ -137,7 +137,7 @@ module.exports = (service, endpoint) => {
-          Submissions.clearDraftSubmissions(form.id)
+          Submissions.softDeleteDraftSubmissions(form.id)
         ]))
         .then(() => Forms.clearUnneededDrafts(form))) // remove drafts made obsolete by new draft
```
Now that this endpoint will no longer result in a new form draft that's ready to be purged, I think we can remove this line. It'd be a nice way to simplify the endpoint. I also think it'd make things clearer if there was only one thing that called `clearUnneededDrafts()` (namely, the purge task).

ktuite replied:
One possible reason to leave it here is when there are no submissions, the draft def can be cleaned up right away. There are some tests in "purging form fields of unneeded drafts" (in `api: /projects/:id/forms (drafts)`) that are checking for these defs to be purged immediately. I could change or remove those tests, though.

matthew-white commented on lib/model/query/submissions.js:
```diff
@@ -121,6 +121,9 @@ const clearDraftSubmissions = (formId) => ({ run }) =>
 const clearDraftSubmissionsForProject = (projectId) => ({ run }) =>
   run(sql`DELETE FROM submissions USING forms WHERE submissions."formId" = forms.id AND forms."projectId" = ${projectId} AND submissions.draft=true`);
 
+const softDeleteDraftSubmissions = (formId) => ({ run }) =>
```
Could we name this `delete` rather than `softDelete`? I can't think of another example offhand where we use the term `softDelete` in the code itself.

ktuite commented:
I've added tests for the 4 main scenarios in the issue, and I've also added some tests showing, experimentally, what it would be like to bring some of these deleted draft submissions back. It gets a little complicated depending on the scenario.

I also had to change the DB submission uniqueness constraint to allow for draft submissions that have been deleted to be able to reuse instance IDs. I am not 100% confident on the uniqueness constraint AND the trigger that exists to check instance ID uniqueness.

matthew-white commented on lib/model/migrations/20241120-01-alter-submission-unique.js:
After discussion on Slack, I think we can revert this migration.

matthew-white commented on lib/model/query/forms.js:
```diff
@@ -485,14 +485,20 @@ const _draftFilter = (form, project) =>
 
-// NOTE: copypasta alert! The following SQL also appears in 20220209-01-purge-unneeded-drafts.js
+// NOTE: copypasta alert! Similar SQL also appears in 20220209-01-purge-unneeded-drafts.js
+// Purges draft form defs that are not referenced by any forms AND have no associated submission defs.
```
```suggestion
// Purges draft form defs that are not referenced by the form as either currentDefId or draftDefId AND have no associated submission defs.
```

ktuite commented on lib/resources/forms.js:
```diff
@@ -162,7 +162,7 @@ module.exports = (service, endpoint) => {
           .then(() => Forms.getByProjectAndXmlFormId(params.projectId, params.id, false, Form.DraftVersion))
           .then(getOrNotFound)
         : resolve(form)))
```
In the case of publishing a draft form with draft submissions... those submissions get soft-deleted but they _could_ be undeleted later. Exclude draft submissions from `Submissions.getDeleted` to prevent this from happening.

ktuite commented on test/integration/api/forms/draft.js:
```diff
@@ -731,118 +731,567 @@ describe('api: /projects/:id/forms (drafts)', () => {
+                body.length.should.equal(2);
+              });
+          }));
         });
```
What happens when `currentDefId` is the same as `draftDefId`?
